dev

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+built_image_name="uaa-bot"
+
+main() {
+  [[ $# -eq 0 ]] && usage "Expected command."
+
+  cd "$(git rev-parse --show-toplevel)"
+
+  command="$1"
+  shift
+
+  case $command in
+    -h)
+      usage
+      ;;
+    black)
+      build_image
+      black "$@"
+      ;;
+    tests)
+      build_image
+      tests "$@"
+      ;;
+    watch-tests)
+      build_image
+      watch-tests "$@"
+      ;;
+    run)
+      build_image
+      run "$@"
+      ;;
+    update-requirements)
+      build_image
+      update-requirements "$@"
+      ;;
+    *)
+      usage "Unknown command: $command"
+      ;;
+  esac
+}
+
+build_image() {
+  # The build-args are used to create a user with the same UID/GID as
+  # yourself. That way, any files written to the $PWD mount will be owned by
+  # you.  This user is embedded in the docker image, so the resulting image
+  # should only be used locally.  These arguments are not used in our
+  # concourse pipeline.
+  echo "Building image..."
+  docker build \
+    --quiet \
+    --file=docker/Dockerfile.dev \
+    --tag=${built_image_name} \
+    --build-arg UID="$(id -u)" \
+    --build-arg GID="$(id -g)" \
+    --build-arg USER="$USER" \
+    .
+}
+
+run_docker() {
```

Comment:
The reason we used docker on the external-domain-broker was because it required coordinating like 3-4 different servers and ports and stuff,  so docker made it possible to do that in both local dev and CI. If you don't need that, I'd consider dropping docker and just using virtualenvs (that's what I'm doing on kibana-cf-auth-proxy)

Reply:
Cool. I've enjoyed developing with this setup and I like the portability of docker for dev work.  I'll update the workflow and pipeline to use virtualenv to simplify though.

Reply:
I'll second @bengerman13's comment about the ease of debugging - I've also had a rough go of it trying to run interactive debuggers with Docker, though I'm curious if that's improved from the last time I tried doing that (this was with Tock about 3 years ago or so).  Did you run into any issues with that, or did you not have a need of it?

I will say, it is cool to see a full setup like this that's working!

Reply:
Pulled out the docker and updated the pipeline for tests. 💯